Move all APIs to use the Twisted-managed connection

jeremycline opened the pull request:
This introduces crochet[0], a small library designed to make it easy to
provide synchronous APIs using asynchronous Twisted APIs. This modifies
the `fedora_messaging.api.consume` and `fedora_messaging.api.publish`
calls to use the Twisted protocol.

There are several advantages here:

1. The synchronous and asynchronous APIs use the same code which means
   there's fewer opportunities for them to differ in behavior and less
   code to maintain.

2. Both APIs didn't really work well because they could not heartbeat in
   the background. This lets them do so because the connection is
   handled in Twisted in a separate crochet-managed thread. While the
   performance improvements are likely modest, it won't lead to so many
   TCP connections being killed by firewalls or the broker.

The downside is it introduces the complexity of threads to the API which
makes it less obvious what is going on.

There are a few minor changes here which slightly change the APIs:

1. Publishing now raises a PublishTimeout when the timeout is reached
   (30 seconds by default).

2. Previously, the Twisted consume API did not validate arguments like
   the synchronous version did, so it now raises a ValueError on invalid
   arguments instead of crashing in some undefined way.

3. Calling publish from the Twisted reactor thread now raises an
   exception instead of blocking the reactor thread.

Although all these are small changes, I think this should be in a 2.0
release. In practice I don't think any of these issues will bite anyone
so upgrading shouldn't be a problem for anyone.

This does not move the CLI back to the synchronous API (yet).

This fixes #171 by unifying the async and sync APIs.

jeremycline commented:
@abompard I squashed the one known issue I had with this PR and I think it's ready for an initial review.

jeremycline commented on fedora_messaging/tests/integration/test_api.py:
```diff
@@ -662,3 +772,307 @@ def test_twisted_consume_update_callback():
+    queues = [
+        {
+            "queue": queue,
+            "auto_delete": True,
```
It looks like it was always true and I just didn't change it when I merged the two integration test sets, I've fixed it and will also pull out the declaration into a fixture.

jeremycline commented on fedora_messaging/tests/integration/test_api.py:
```diff
@@ -662,3 +772,307 @@ def test_twisted_consume_update_callback():
+    client = yield serv.getFactory().whenConnected()
+    channel = yield client._allocate_channel()
+    assert channel._delivery_confirmation is True
+    serv.stopService()
```
startService doesn't return a deferred, but you're right about stopService. Fixed.

jeremycline commented:
Okay, I've addressed the PR comments. If you're happy with the general shape of things, I'll resolve the merge conflicts and see about adding a bit more test coverage and we should be ready. I already packaged crochet for Fedora and we can built it in the infra repository for epel7.

abompard commented:
Yeah I'm happy with the state of things, and happy to have fewer codepaths :-) Thanks!

jeremycline commented:
Okay, I've rebased and added a couple more tests. The lack of coverage is in a compatibility block for EL7 which is difficult to test in CI, but would be covered by tests run in EL7. I'm inclined to leave them as-is until we drop support for that. The other spot I couldn't easily get coverage was in the publish code that handled connection errors during the publish. I could mock it all out and test that, but I'd prefer a test against an actual broker, I've just not come up with a neat way to do it yet.
